ksql-execution/src/main/java/io/confluent/ksql/execution/plan/StreamSource.java

```diff
@@ -14,35 +14,70 @@
 
 package io.confluent.ksql.execution.plan;
 
+import com.google.common.annotations.VisibleForTesting;
 import com.google.errorprone.annotations.Immutable;
 import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.execution.builder.KsqlQueryBuilder;
+import io.confluent.ksql.execution.context.QueryContext;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.schema.ksql.PhysicalSchema;
+import io.confluent.ksql.serde.KeyFormat;
+import io.confluent.ksql.serde.KeySerde;
+import io.confluent.ksql.serde.SerdeOption;
+import io.confluent.ksql.serde.ValueFormat;
+import io.confluent.ksql.util.Pair;
 import io.confluent.ksql.util.timestamp.TimestampExtractionPolicy;
 import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
-import org.apache.kafka.streams.StreamsBuilder;
-import org.apache.kafka.streams.Topology;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.BiFunction;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.streams.Topology.AutoOffsetReset;
+import org.apache.kafka.streams.kstream.Consumed;
 import org.apache.kafka.streams.kstream.KStream;
+import org.apache.kafka.streams.kstream.KsqlConsumed;
+import org.apache.kafka.streams.kstream.ValueMapperWithKey;
+import org.apache.kafka.streams.kstream.ValueTransformer;
+import org.apache.kafka.streams.kstream.ValueTransformerSupplier;
+import org.apache.kafka.streams.kstream.Window;
+import org.apache.kafka.streams.kstream.Windowed;
+import org.apache.kafka.streams.kstream.internals.SessionWindow;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.TimestampExtractor;
 
 @Immutable
 public class StreamSource<K> implements ExecutionStep<KStream<K, GenericRow>> {
   private final ExecutionStepProperties properties;
   private final String topicName;
   private final Formats formats;
   private final TimestampExtractionPolicy timestampPolicy;
-  private final Topology.AutoOffsetReset offsetReset;
+  private final int timestampIndex;
+  private final Optional<AutoOffsetReset> offsetReset;
+  private final LogicalSchema sourceSchema;
+  // These fields are not serialized. They are computed when deserialized
+  private final Functions<K> functions;
 
-  public StreamSource(
+  @VisibleForTesting
+  StreamSource(
```

Comment:
> The third could be moved into a separate class. This can again be some kind of builder/factory class specific t this step, and distinct from the class used for the second step, or could be some KS specific implementation of a PhsycialNodeVisitor that coverts the physical plan to a KS implementation.

I definitely agree here, already have a follow-up in this direction in progress